components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,262 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use app_units::Au;
+use dom_struct::dom_struct;
+use euclid::default::Rect;
+use js::rust::HandleObject;
+
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::ResizeObserverBinding::{
+    ResizeObserverBoxOptions, ResizeObserverCallback, ResizeObserverMethods, ResizeObserverOptions,
+};
+use crate::dom::bindings::codegen::Bindings::WindowBinding::WindowMethods;
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::domrectreadonly::DOMRectReadOnly;
+use crate::dom::element::Element;
+use crate::dom::node::{window_from_node, Node};
+use crate::dom::resizeobserverentry::ResizeObserverEntry;
+use crate::dom::resizeobserversize::{ResizeObserverSize, ResizeObserverSizeImpl};
+use crate::dom::window::Window;
+
+/// <https://drafts.csswg.org/resize-observer/#calculate-depth-for-node>
+#[derive(Debug, Default, PartialEq, PartialOrd)]
+pub struct ResizeObservationDepth(usize);
+
+impl ResizeObservationDepth {
+    pub fn max() -> ResizeObservationDepth {
+        ResizeObservationDepth(usize::MAX)
+    }
+}
+
+/// <https://drafts.csswg.org/resize-observer/#resize-observer-slots>
+/// See `ObservationState` for active and skipped observation targets.
+#[dom_struct]
+pub struct ResizeObserver {
+    reflector_: Reflector,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-callback-slot>
+    #[ignore_malloc_size_of = "Rc are hard"]
+    callback: Rc<ResizeObserverCallback>,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-observationtargets-slot>
+    observation_targets: DomRefCell<Vec<(ResizeObservation, Dom<Element>)>>,
+}
+
+impl ResizeObserver {
+    pub fn new_inherited(callback: Rc<ResizeObserverCallback>) -> ResizeObserver {
+        ResizeObserver {
+            reflector_: Reflector::new(),
+            callback,
+            observation_targets: Default::default(),
+        }
+    }
+
+    fn new(
+        window: &Window,
+        proto: Option<HandleObject>,
+        callback: Rc<ResizeObserverCallback>,
+    ) -> DomRoot<ResizeObserver> {
+        let observer = Box::new(ResizeObserver::new_inherited(callback));
+        reflect_dom_object_with_proto(observer, window, proto)
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-resizeobserver>
+    #[allow(non_snake_case)]
+    pub fn Constructor(
+        window: &Window,
+        proto: Option<HandleObject>,
+        callback: Rc<ResizeObserverCallback>,
+    ) -> DomRoot<ResizeObserver> {
+        let rooted_observer = ResizeObserver::new(window, proto, callback);
+        let document = window.Document();
+        document.add_resize_observer(&rooted_observer);
```

Comment:
This is per spec, but has memory problems: https://bugzilla.mozilla.org/show_bug.cgi?id=1596992

I don't remember how Blink managed to avoid that, but I think you should either do the same or disobey the spec like Gecko and WebKit do: https://github.com/w3c/csswg-drafts/issues/4518

Reply:
Thanks for pointing it out, there is a note over at https://github.com/servo/servo/pull/31108/files#diff-b6bdaf97b920500ab94dea9827cda04ff8136b0f23bd0c515d9195c6be31a820R439

and I'll add the info you mention to it...